File: tests/aff4_image_test.cc
/*
Copyright 2015 Google Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License.  You may obtain a copy of the
License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.  See the License for the
specific language governing permissions and limitations under the License.
*/
#include <gtest/gtest.h>
#include "aff4/libaff4.h"
#include <unistd.h>
#include <glog/logging.h>
#include "utils.h"

namespace aff4 {


class AFF4ImageTest: public ::testing::Test {
 protected:
  std::string filename = "/tmp/aff4_test.zip";
  std::string image_name = "image.dd";
  URN volume_urn;
  URN image_urn;
  URN image_urn_2;
  URN image_urn_stream;

  // Remove the file on teardown.
  virtual void TearDown() {
      unlink(filename.c_str());
  }

  // Create an AFF4Image stream with some data in it.
  virtual void SetUp() {
    MemoryDataStore resolver;

    unlink(filename.c_str());

    AFF4Flusher<AFF4Stream> file;
    AFF4Flusher<ZipFile> zip;
    EXPECT_OK(NewFileBackedObject(&resolver, filename, "truncate", file));
    EXPECT_OK(ZipFile::NewZipFile(&resolver, std::move(file), zip));

    volume_urn = zip->urn;

    {
      image_urn = zip->urn.Append(image_name);

      AFF4Flusher<AFF4Image> image;
      EXPECT_OK(AFF4Image::NewAFF4Image(
                    &resolver, image_urn, zip.get(), image));

      // For testing - rediculously small chunks. This will create many bevies.
      image->chunk_size = 10;
      image->chunks_per_segment = 3;

      for (int i = 0; i < 100; i++) {
        image->sprintf("Hello world %02d!", i);
      }
    }

    // Now test snappy compression in images.
    {
      image_urn_2 = image_urn.Append("2");

      AFF4Flusher<AFF4Image> image_2;
      EXPECT_OK(AFF4Image::NewAFF4Image(
                    &resolver, image_urn_2, zip.get(), image_2));

      // Make the second image use snappy for compression.
      image_2->compression = AFF4_IMAGE_COMPRESSION_ENUM_SNAPPY;
      image_2->Write("This is a test");
    }

    // Now test the streaming API.
    {
      std::unique_ptr<AFF4Stream> test_stream = StringIO::NewStringIO();
      test_stream->Write("This is a test");
      test_stream->Seek(0, SEEK_SET);

      image_urn_stream = zip->urn.Append(image_name).Append("stream");

      AFF4Flusher<AFF4Image> image;
      EXPECT_OK(AFF4Image::NewAFF4Image(
                    &resolver, image_urn_stream, zip.get(), image));

      // For testing - rediculously small chunks. This will create many bevies.
      image->chunk_size = 10;
      image->chunks_per_segment = 3;
      image->compression = AFF4_IMAGE_COMPRESSION_ENUM_SNAPPY;

      image->WriteStream(test_stream.get());
    }
  }
};

TEST_F(AFF4ImageTest, OpenImageByURN) {
  MemoryDataStore resolver;

  AFF4Flusher<AFF4Stream> file;
  AFF4Flusher<AFF4Volume> zip;
  EXPECT_OK(NewFileBackedObject(&resolver, filename, "read", file));
  EXPECT_OK(ZipFile::OpenZipFile(&resolver, std::move(file), zip));

  VolumeGroup volumes(&resolver);
  volumes.AddVolume(std::move(zip));

  AFF4Flusher<AFF4Image> image;
  EXPECT_OK(AFF4Image::OpenAFF4Image(
                &resolver, image_urn, &volumes, image));

  // Ensure the newly opened image has the correct parameters.
  EXPECT_EQ(image->chunk_size, 10);
  EXPECT_EQ(image->chunks_per_segment, 3);

  EXPECT_STREQ(
      "Hello world 00!Hello world 01!Hello world 02!Hello world 03!"
      "Hello world 04!Hello world 05!Hello worl",
      image->Read(100).c_str());

  EXPECT_EQ(1500, image->Size());
}


TEST_F(AFF4ImageTest, TestAFF4ImageStream) {
  MemoryDataStore resolver;

  AFF4Flusher<AFF4Stream> file;
  AFF4Flusher<AFF4Volume> zip;
  EXPECT_OK(NewFileBackedObject(&resolver, filename, "read", file));
  EXPECT_OK(ZipFile::OpenZipFile(&resolver, std::move(file), zip));

  VolumeGroup volumes(&resolver);
  volumes.AddVolume(std::move(zip));

  AFF4Flusher<AFF4Image> image;
  EXPECT_OK(AFF4Image::OpenAFF4Image(
                &resolver, image_urn, &volumes, image));

  std::unique_ptr<StringIO> stream_copy = StringIO::NewStringIO();
  for (int i = 0; i < 100; i++) {
    stream_copy->sprintf("Hello world %02d!", i);
  }

  for (int i = 0; i < 1500; i += 25) {
    image->Seek(i, SEEK_SET);
    stream_copy->Seek(i, SEEK_SET);

    // Randomly read buffers in the image to ensure Seek/Read works.
    std::string read_data = image->Read(13);
    std::string expected_data = stream_copy->Read(13);

    resolver.logger->info("Expected: {}", expected_data);
    resolver.logger->info("Read: {}", read_data);

    EXPECT_STREQ(expected_data.c_str(), read_data.c_str());
  }

  {
    // Now test snappy decompression.
    AFF4Flusher<AFF4Image> image_2;
    EXPECT_OK(AFF4Image::OpenAFF4Image(
                  &resolver, image_urn_2, &volumes, image_2));

    URN compression_urn;
    EXPECT_OK(resolver.Get(image_2->urn, AFF4_IMAGE_COMPRESSION, compression_urn));

    EXPECT_STREQ(compression_urn.SerializeToString().c_str(),
                 AFF4_IMAGE_COMPRESSION_SNAPPY);

    std::string data = image_2->Read(100);
    EXPECT_STREQ(data.c_str(), "This is a test");
  }

  // Now test streaming interface.
  {
      AFF4Flusher<AFF4Image> image_stream;
      EXPECT_OK(AFF4Image::OpenAFF4Image(
                    &resolver, image_urn_stream, &volumes, image_stream));

      URN compression_urn;
      EXPECT_OK(resolver.Get(image_stream->urn, AFF4_IMAGE_COMPRESSION,
                             compression_urn));

      EXPECT_STREQ(compression_urn.SerializeToString().c_str(),
                   AFF4_IMAGE_COMPRESSION_SNAPPY);

      std::string data = image_stream->Read(100);
      EXPECT_STREQ(data.c_str(), "This is a test");
  }
}


} // namespace aff4

File: pyaff4/pyaff4/aff4image_test.py
# Copyright 2014 Google Inc. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

import os
import StringIO
import unittest

from pyaff4 import aff4_image
from pyaff4 import data_store
from pyaff4 import lexicon
from pyaff4 import rdfvalue
from pyaff4 import zip

from pyaff4 import plugins


class AFF4ImageTest(unittest.TestCase):
    filename = "/tmp/aff4_test.zip"
    image_name = "image.dd"

    def tearDown(self):
        try:
            os.unlink(self.filename)
        except (IOError, OSError):
            pass

    def setUp(self):
        with data_store.MemoryDataStore() as resolver:
            resolver.Set(self.filename, lexicon.AFF4_STREAM_WRITE_MODE,
                         rdfvalue.XSDString("truncate"))

            with zip.ZipFile.NewZipFile(resolver, self.filename) as zip_file:
                self.volume_urn = zip_file.urn
                image_urn = self.volume_urn.Append(self.image_name)

                # Use default compression.
                with aff4_image.AFF4Image.NewAFF4Image(
                    resolver, image_urn, self.volume_urn) as image:
                    image.chunk_size = 10
                    image.chunks_per_segment = 3

                    for i in range(100):
                        image.Write("Hello world %02d!" % i)

                    self.image_urn = image.urn

                # Write a snappy compressed image.
                self.image_urn_2 = self.image_urn.Append("2")
                with aff4_image.AFF4Image.NewAFF4Image(
                    resolver, self.image_urn_2, self.volume_urn) as image_2:
                    image_2.compression = lexicon.AFF4_IMAGE_COMPRESSION_SNAPPY
                    image_2.Write("This is a test")

                # Use streaming API to write image.
                self.image_urn_3 = self.image_urn.Append("3")
                with aff4_image.AFF4Image.NewAFF4Image(
                    resolver, self.image_urn_3, self.volume_urn) as image:
                    image.chunk_size = 10
                    image.chunks_per_segment = 3
                    stream = StringIO.StringIO()
                    for i in range(100):
                        stream.write("Hello world %02d!" % i)

                    stream.seek(0)
                    image.WriteStream(stream)

    def testOpenImageByURN(self):
        resolver = data_store.MemoryDataStore()

        # This is required in order to load and parse metadata from this volume
        # into a fresh empty resolver.
        with zip.ZipFile.NewZipFile(resolver, self.filename) as zip_file:
            image_urn = zip_file.urn.Append(self.image_name)

        with resolver.AFF4FactoryOpen(image_urn) as image:
            self.assertEquals(image.chunk_size, 10)
            self.assertEquals(image.chunks_per_segment, 3)
            self.assertEquals(
                "Hello world 00!Hello world 01!Hello world 02!Hello world 03!"
                "Hello world 04!Hello world 05!Hello worl",
                image.Read(100))

            self.assertEquals(1500, image.Size())

        # Now test snappy decompression.
        with resolver.AFF4FactoryOpen(self.image_urn_2) as image_2:
            self.assertEquals(
                resolver.Get(image_2.urn, lexicon.AFF4_IMAGE_COMPRESSION),
                lexicon.AFF4_IMAGE_COMPRESSION_SNAPPY)

            data = image_2.Read(100)
            self.assertEquals(data, "This is a test")

        # Now test streaming API image.
        with resolver.AFF4FactoryOpen(self.image_urn_3) as image_3:
            self.assertEquals(image_3.chunk_size, 10)
            self.assertEquals(image_3.chunks_per_segment, 3)
            self.assertEquals(
                "Hello world 00!Hello world 01!Hello world 02!Hello world 03!"
                "Hello world 04!Hello world 05!Hello worl",
                image_3.Read(100))

if __name__ == '__main__':
    #logging.getLogger().setLevel(logging.DEBUG)
    unittest.main()
